### PR TITLE
Show answer button config

### PR DESCRIPTION
**Pull request by jlikhuva**
@caesar2164 @stvstnfrd @caseylitton  PR for new Studio Feature requested by Jeff Ullman. Allow course creators to specify that the show answer button only shows up after the student makes at least # attempts.
@jspayd @potsui

When testing, ensure that you are logged in as a student to see the effects of the question settings.

### CODE REVIEW

File: common/lib/xmodule/xmodule/capa_base.py [hunk 1]
```diff
@@ -137,7 +137,15 @@ class CapaFields(object):
             {"display_name": _("Finished"), "value": SHOWANSWER.FINISHED},
             {"display_name": _("Correct or Past Due"), "value": SHOWANSWER.CORRECT_OR_PAST_DUE},
             {"display_name": _("Past Due"), "value": SHOWANSWER.PAST_DUE},
-            {"display_name": _("Never"), "value": SHOWANSWER.NEVER}]
+            {"display_name": _("Never"), "value": SHOWANSWER.NEVER},
+            {"display_name": _("After # Attempts"), "value": SHOWANSWER.AFTER_ATTEMPTS}]
```
Comment (caesar2164) [inline common/lib/xmodule/xmodule/capa_base.py]: thanks @jlikhuva ! 😄

File: common/lib/xmodule/xmodule/capa_base.py [hunk 2]
```diff
@@ -899,6 +907,8 @@ def answer_available(self):
             return self.is_correct() or self.is_past_due()
         elif self.showanswer == SHOWANSWER.PAST_DUE:
             return self.is_past_due()
+        elif self.showanswer == SHOWANSWER.AFTER_ATTEMPTS:
+            return self.attempts >= self.attempts_before_showanswer_button
```
Comment (caesar2164) [inline common/lib/xmodule/xmodule/capa_base.py]: can we add a line inside this `elif` like so?
```python
if self.attempts_before_showanswer_button >= self.max_attempts:
    self.attempts_before_showanswer_button = self.max_attempts
```

File: common/lib/xmodule/xmodule/tests/test_capa_module.py [hunk 3]
```diff
@@ -361,6 +361,49 @@ def test_showanswer_past_due(self):
...
+
+        # Can see even though grace period hasn't expired as long as
+        # User has already attempted for the required number of times
+        still_in_grace = CapaFactory.create(showanswer='after_attempts',
+                                            after_attempts="3",
+                                            attempts="4",
+                                            due=self.yesterday_str,
+                                            graceperiod=self.two_day_delta_str)
+        self.assertTrue(still_in_grace.answer_available())
+
```
Comment (caesar2164) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: please add a test that checks that your code doesn't fall over if `after_attempts` > `max_attempts`

**Comment by caseylitton** [comment]
I checked the functionality on stage for 'attempted' and 'answered' and cannot get the 'show answer' button to disappear unless I select 'never'.  I may be missing something on the setup in stage.  

Please provide some instructions for setting up the 'after_x_attempts' type of problem in studio settings.

File: common/lib/xmodule/xmodule/capa_base.py [hunk 4]
```diff
@@ -137,7 +137,15 @@ class CapaFields(object):
             {"display_name": _("Finished"), "value": SHOWANSWER.FINISHED},
             {"display_name": _("Correct or Past Due"), "value": SHOWANSWER.CORRECT_OR_PAST_DUE},
             {"display_name": _("Past Due"), "value": SHOWANSWER.PAST_DUE},
-            {"display_name": _("Never"), "value": SHOWANSWER.NEVER}]
+            {"display_name": _("Never"), "value": SHOWANSWER.NEVER},
+            {"display_name": _("After # Attempts"), "value": SHOWANSWER.AFTER_ATTEMPTS}]
+    )
+    attempts_before_showanswer_button = Integer(
+        display_name=_("Show Answer - Attempts"),
```
Comment (caseylitton) [inline common/lib/xmodule/xmodule/capa_base.py]: Similar to the pound sign, should 'Show Answer - Attempts' be 'Show Answer After Attempts' ?

File: common/lib/xmodule/xmodule/tests/test_capa_module.py (hunk 3 above)
Comment (caseylitton) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: Please check the tests locally before resubmitting

paver test_system -t common/lib/xmodule/xmodule/tests/test_capa_module.py:CapaModuleTest

Comment (caseylitton) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: Also after_attempts could be zero.

File: common/lib/xmodule/xmodule/capa_base.py (hunk 1 above)
Comment (caesar2164) [inline common/lib/xmodule/xmodule/capa_base.py]: I guess it could just be "After Attempts"?

File: common/lib/xmodule/xmodule/tests/test_capa_module.py [hunk 5]
```diff
@@ -361,6 +361,65 @@ def test_showanswer_past_due(self):
...
+        """
+        With showanswer="after_attempts",
+        the show answer button should
+        only appear after the user has
+        attempted the problem(i.e clicked the
+        submit button) for the requisite number of times.
+
+        Once the button appears, it does not disappear again.
+        """
+        # Cannot see answer because user has not attempted
```
Comment (caesar2164) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: can you put a line before each comment so that each "test" is visually separated?

File: common/lib/xmodule/xmodule/tests/test_capa_module.py [hunk 6]
```diff
@@ -361,6 +361,65 @@ def test_showanswer_past_due(self):
...
+        # answering question for the required number of
+        # times.
+        before_attempts_with_max = CapaFactory.create(showanswer="after_attempts",
+                                                      attempts="2",
+                                                      attempts_before_showanswer_button="3",
+                                                      max_attempts="5")
+        self.assertFalse(before_attempts_with_max.answer_available())
+        # can't see answer even when max_attempts is not set.
```
Comment (caesar2164) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: line above here

File: common/lib/xmodule/xmodule/tests/test_capa_module.py [hunk 7]
```diff
@@ -361,6 +361,65 @@ def test_showanswer_past_due(self):
...
+        before_attempts_no_max = CapaFactory.create(showanswer="after_attempts",
+                                                    attempts="2",
+                                                    attempts_before_showanswer_button="3")
+        self.assertFalse(before_attempts_no_max.answer_available())
+        # can see answer even after all attempts are used up,
```
Comment (caesar2164) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: line above

File: common/lib/xmodule/xmodule/tests/test_capa_module.py [hunk 8]
```diff
@@ -361,6 +361,65 @@ def test_showanswer_past_due(self):
...
+        # even with the due date in the future
+        used_all_attempts = CapaFactory.create(showanswer='after_attempts',
+                                               attempts_before_showanswer_button="2",
+                                               max_attempts="3",
+                                               attempts="3",
+                                               due=self.tomorrow_str)
+        self.assertTrue(used_all_attempts.answer_available())
+        # can see after due date
```
Comment (caesar2164) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: etc...

File: common/lib/xmodule/xmodule/tests/test_capa_module.py (hunk 5 above)
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: These should actually just be separate test cases.

Each test should generally only assert one thing.
With proper naming, the comments won't even be necessary.
Or if it makes sense to keep the comments, they should be docstrings to the individual methods.

File: common/lib/xmodule/xmodule/tests/test_capa_module.py [hunk 9]
```diff
@@ -361,6 +361,65 @@ def test_showanswer_past_due(self):
...
+        only appear after the user has
+        attempted the problem(i.e clicked the
+        submit button) for the requisite number of times.
+
+        Once the button appears, it does not disappear again.
+        """
+        # Cannot see answer because user has not attempted
+        # answering question for the required number of
+        # times.
+        before_attempts_with_max = CapaFactory.create(showanswer="after_attempts",
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: Multiple lines like this shouldn't be aligned under the parenthesis;
please follow the formatting specified here:
https://github.com/stvstnfrd/edx-platform/blob/7c709518474446d00c782b1e5e37c3137f45d9d6/openedx/stanford/style/python/collections-multiline.markdown#explanation

File: common/lib/xmodule/xmodule/capa_base.py [hunk 10]
```diff
@@ -105,6 +105,7 @@ class CapaFields(object):
         scope=Scope.user_state)
     max_attempts = Integer(
         display_name=_("Maximum Attempts"),
+        #default=float('inf'),
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/capa_base.py]: Remove commented code.

https://github.com/stvstnfrd/edx-platform/blob/273a75a7389988fe84ddd014c9e993c564af834c/openedx/stanford/style/python/comments-code.markdown

File: common/lib/xmodule/xmodule/capa_base.py (hunk 1 above)
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/capa_base.py]: I agree with Casey re: `#`; I think the objection here is the symbol chosen (not obvious/universal), not the idea.

I'd prefer something more explicit:

```python
{"display_name": _("After Some Number of Attempts"), "value": SHOWANSWER.AFTER_SOME_NUMBER_OF_ATTEMPTS}
```

"After Attempts" could imply "After _all_ Attempts" or "After _any_ Attempts".
We want it to be more clear that this is after an an _arbitrary_ number of attempts,
not after _all_ or _any_.

File: common/lib/xmodule/xmodule/capa_base.py [hunk 11]
```diff
@@ -137,7 +138,15 @@ class CapaFields(object):
...
-            {"display_name": _("Never"), "value": SHOWANSWER.NEVER}]
+            {"display_name": _("Never"), "value": SHOWANSWER.NEVER},
+            {"display_name": _("After # Attempts"), "value": SHOWANSWER.AFTER_ATTEMPTS}]
+    )
+    attempts_before_showanswer_button = Integer(
+        display_name=_("Show Answer After Attempts"),
+        help=_("Number of times the student must attempt answering the question before the Show Answer button appears."),
+        values={"min": 0},
+        default=0,
+        scope=Scope.settings
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/capa_base.py]: Add a trailing comma.

https://github.com/stvstnfrd/edx-platform/blob/273a75a7389988fe84ddd014c9e993c564af834c/openedx/stanford/style/python/collections-trailing-comma.markdown

File: common/lib/xmodule/xmodule/capa_base.py [hunk 12]
```diff
@@ -899,9 +908,15 @@ def answer_available(self):
...
+        elif self.showanswer == SHOWANSWER.AFTER_ATTEMPTS:
+            # self.attempts_before_showanswer_button is read_only, to
+            # work with its modified value, we make a copy.
+            required_attempts = self.attempts_before_showanswer_button
+            if self.max_attempts and required_attempts >= self.max_attempts:
+                required_attempts = self.max_attempts
+            return self.attempts >= required_attempts
         elif self.showanswer == SHOWANSWER.ALWAYS:
             return True
-
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/capa_base.py]: This is an unnecessary change, so we should revert it.

We want to keep PRs focused and as minimal as possible,
so we should only change lines that apply to the feature at hand.

Unrelated cleanup can be addressed in a separate PR.

File: common/lib/xmodule/xmodule/tests/test_capa_module.py [hunk 13]
```diff
@@ -361,6 +361,65 @@ def test_showanswer_past_due(self):
...
+        still_in_grace = CapaFactory.create(showanswer='after_attempts',
+                                            after_attempts="3",
+                                            attempts="4",
+                                            due=self.yesterday_str,
+                                            graceperiod=self.two_day_delta_str)
+        self.assertTrue(still_in_grace.answer_available())
+        # Ensure that if attempts_before_showanswer_button > max_attempts, the button shows up
+        # after all attempts are used up, i.e after_attempts becomes max_attempts
+        large_after_attempts = CapaFactory.create(showanswer='after_attempts',
+                                                  attempts_before_showanswer_button="5",
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: Here and throughout, we're switching between single and double quotes.

For consistency, let's just use single quotes.

Note: this comment only applies to the new code you've added here;
don't change the existing code.

File: common/lib/xmodule/xmodule/capa_base.py [hunk 14]
```diff
@@ -899,9 +908,15 @@ def answer_available(self):
             return self.is_correct() or self.is_past_due()
         elif self.showanswer == SHOWANSWER.PAST_DUE:
             return self.is_past_due()
+        elif self.showanswer == SHOWANSWER.AFTER_ATTEMPTS:
+            # self.attempts_before_showanswer_button is read_only, to
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/capa_base.py]: This comment may not be necessary.

File: common/lib/xmodule/xmodule/capa_base.py [hunk 15]
```diff
@@ -137,7 +137,15 @@ class CapaFields(object):
             {"display_name": _("Finished"), "value": SHOWANSWER.FINISHED},
             {"display_name": _("Correct or Past Due"), "value": SHOWANSWER.CORRECT_OR_PAST_DUE},
             {"display_name": _("Past Due"), "value": SHOWANSWER.PAST_DUE},
-            {"display_name": _("Never"), "value": SHOWANSWER.NEVER}]
+            {"display_name": _("Never"), "value": SHOWANSWER.NEVER},
+            {"display_name": _("After Some Number of Attempts"), "value": SHOWANSWER.AFTER_SOME_NUMBER_OF_ATTEMPTS}]
```
Comment (caesar2164) [inline common/lib/xmodule/xmodule/capa_base.py]: needs a trailing comma and the square brace should drop to next line

**Comment by jlikhuva** [comment]
@stvstnfrd @caesar2164  Refactored the unit test in `common/lib/xmodule/xmodule/tests/test_capa_module.py` into a bunch of smaller independent unit tests.

Also, Changes the label on the new entry boxes.

Will incorporate @caesar2164's  latest review in next commit.

**Comment by caesar2164** [comment]
test this please

File: common/lib/xmodule/xmodule/tests/test_capa_module.py [hunk 16]
```diff
@@ -361,6 +361,91 @@ def test_showanswer_past_due(self):
                                             graceperiod=self.two_day_delta_str)
         self.assertFalse(still_in_grace.answer_available())
 
+    def test_showanswer_after_attempts_with_max(self):
```
Comment (caseylitton) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: All tests should have comments and the first line of the comment should be meaningful.

The first line of the comment is printed when the test fails so that line should ideally be a compete sentence or idea.  It should differentiate the test from other tests like it but also be similar.  You can use subsequent lines to further explain the test if needed.  You have 50 or 80 characters on the first line so use them. 

Below are possible comment examples for the first three tests.  When I make a group of tests I will try to be consistent with the comments and function names.  Maybe you'll have a better idea for having similar comments since you're more familiar with the test group.

Show answer button should NOT be available when attempts is less than attempts before.
Show answer button should NOT be available when attempts is less than attempts before.
Show answer button should be available when attempts > attempts_before.

File: common/lib/xmodule/xmodule/tests/test_capa_module.py [hunk 17]
```diff
@@ -361,6 +361,91 @@ def test_showanswer_past_due(self):
...
+        )
+        self.assertTrue(large_after_attempts.answer_available())
+
+    def test_showanswer_after_attempts_zero(self):
+        '''
+        If attempts_before_showanswer_button is 0,
+        then the show_answer button should always
+        be visible.
+        '''
+        zero_after_attempts = CapaFactory.create(
```
Comment (caseylitton) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: Other people may disagree but I will use the same variable names in group of tests like these.    

The function name and comments should tell people what is going on.  We don't gain much in this scenario by having descriptive variable names.  It made sense when all the tests were in one function however.

It's extra development work with little to no gain in understanding.   I think it's actually more confusing.

Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: Agree; I'll usually use the general term here, like `problem` or some such, that way similar lines look the same:

```python
problem = CapaFactor.create(
    # ...
)
```

This makes subsequent lines even easier to comprehend:

```python
assert problem.is_valid()
# vs
assert zero_after_attempts.is_valid()
```

File: common/lib/xmodule/xmodule/tests/test_capa_module.py (hunk 16 above)
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: Admittedly, this is one area where I'm willing to be _slightly_ more lenient...
_if_ _and_ _only_ _if_ the test name itself is sufficiently expressive.

When there is no docstring, the function name can be used instead;
_Sometimes_ that suffices.

Consider:

```python
def test_show_answer_after_max_attempts(self):
    """
    Test that show answer functions properly after the max attempts has been reached
    """
    pass
```

Here, the comment just rewords the test name, so I would argue the comment is not necessary.

On the other hand:

```python
def test_show_answer_after_max_attempts_but_only_if_we_have_configured_some_certain_values_which_may_affect_this(self):
    pass
```

That's far too wordy for a method name, so the method name should be shortened and the special case explained via docstring.

Here though, I agree with @caseylitton; this could be expressed a bit more clearly.

**Comment by jlikhuva** [comment]
@caseylitton @stvstnfrd @caesar2164 

Updated the documentation on the Unit tests.

File: common/lib/xmodule/xmodule/tests/test_capa_module.py [hunk 18]
```diff
@@ -361,6 +361,108 @@ def test_showanswer_past_due(self):
...
+    def test_showanswer_after_attempts_zero(self):
+        '''
+        Button should always be visible if required min attempts = 0.
+        If attempts_before_showanswer_button = 0, then the show answer
+        button should be visible at all times.
+        '''
+        problem = CapaFactory.create(
+            showanswer='after_attempts',
+            attempts_before_showanswer_button='0',
+            attempts='0'
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: Add trailing commas here and throughout.

https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/style/python/collections-trailing-comma.markdown

File: common/lib/xmodule/xmodule/tests/test_capa_module.py [hunk 19]
```diff
@@ -361,6 +361,108 @@ def test_showanswer_past_due(self):
...
+    def test_showanswer_after_attempts_large(self):
+        '''
+        If required attempts > max attempts then required attempts = max attempts.
+        Ensure that if attempts_before_showanswer_button > max_attempts,
+        the button should show up after all attempts are used up,
+        i.e after_attempts falls back to max_attempts
+        '''
+        problem = CapaFactory.create(
+            showanswer='after_attempts',
+            attempts_before_showanswer_button="5",
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: We should use single quotes consistently, here and throughout.
Note: do not changes the quotes on existing code, only for the lines you're adding/modifying here.

File: common/lib/xmodule/xmodule/tests/test_capa_module.py [hunk 20]
```diff
@@ -361,6 +361,108 @@ def test_showanswer_past_due(self):
                                             graceperiod=self.two_day_delta_str)
         self.assertFalse(still_in_grace.answer_available())
 
+    def test_showanswer_after_attempts_with_max(self):
+        '''
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: Use double quotes for docstrings.

https://github.com/Stanford-Online/edx-platform/blob/master/openedx/stanford/style/python/docstrings.markdown

File: common/lib/xmodule/xmodule/tests/test_capa_module.py [hunk 21]
```diff
@@ -361,6 +361,108 @@ def test_showanswer_past_due(self):
...
+        Button should not be visible when attempts < required attempts.
```
Comment (stvstnfrd) [inline common/lib/xmodule/xmodule/tests/test_capa_module.py]: Per f2f, force one of these tests to fail, so that we can see how it formats this message/title.
We may need a newline after the first line of text.
Actually, I think best practices actually recommend that, so let's go ahead and do it regardless.

```python
def foo:
    """
    Button should not be visible when attempts < required attempts.

    Even with max attempts set, the show answer button should only
    show up after the user has attempted answering the question for
    the requisite number of times, i.e `attempts_before_showanswer_button`
    """
```

**Comment by jlikhuva** [comment]
@stvstnfrd  Incorporated your comments in the latest commit

**Comment by stvstnfrd** [comment]
This looks good @jlikhuva !
Let's go ahead an open up a PR to upstream with these changes.

**Comment by caseylitton** [comment]
+1 make sure to squash.
